File: 73.minimum_window_substring.ts
function minWindow1(s: string, t: string): string {
	let result: string = ""
	const sMap: Record<string, number> = {}
	const tMap: Record<string, number> = {}
	let left: number = 0
	let right: number = 0

	for (let i: number = 0; i < t.length; i++) {
		if (tMap[t[i]]) {
			tMap[t[i]] = tMap[t[i]] + 1
		} else {
			tMap[t[i]] = 1
		}
	}

	while (right <= s.length) {
		if (check(sMap, tMap)) {
			if (right - left < result.length || result == "") {
				result = s.substring(left, right)
			}
			sMap[s[left]] = sMap[s[left]] - 1
			left++
		} else {
			if (right < s.length) {
				if (sMap[s[right]]) {
					sMap[s[right]] = sMap[s[right]] + 1
				} else {
					sMap[s[right]] = 1
				}
			}
			right++
		}
	}

	return result
};

function check(sMap: Record<string, number>, tMap: Record<string, number>): boolean {
	for (const [key, value] of Object.entries(tMap)) {
		if (!sMap[key]) {
			return false
		} else if (sMap[key] < tMap[key]) {
			return false
		}
	}
	return true
}

function minWindow2(s: string, t: string): string {
	const map = {};
	for (let i = 0; i < t.length; i++) {
		const curStr = t[i];
		if (!map[curStr]) {
			map[curStr] = 1;
		} else {
			map[curStr]++;
		}
	}

	let start = 0;
	let end = 0;
	let currWinLength = 0;
	let minResLength = -1;
	let res = '';

	while (end < s.length) {
		const curStr = s[end];
		if (!map[curStr]) {
			map[curStr]--;
		} else {
			map[curStr]--;
		}
		if (map[curStr] >= 0) currWinLength++;
		while (currWinLength === t.length) {
			if (minResLength === -1 || minResLength > end - start + 1) {
				res = s.slice(start, end + 1);
				minResLength = end - start + 1;
			}
			const curStr = s[start];
			map[curStr]++;
			if (map[curStr] >= 1) currWinLength--;
			start++;
		}
		end++;
	}
	return res;
};


minWindow2("ADOBECODEBANC", "ABC")

File: 73.minimum_window_substring.go
package main

/*
Given two strings s and t of lengths m and n respectively, return the minimum window
substring
 of s such that every character in t (including duplicates) is included in the window. If there is no such substring, return the empty string "".

The testcases will be generated such that the answer is unique.



Example 1:

Input: s = "ADOBECODEBANC", t = "ABC"
Output: "BANC"
Explanation: The minimum window substring "BANC" includes 'A', 'B', and 'C' from string t.
Example 2:

Input: s = "a", t = "a"
Output: "a"
Explanation: The entire string s is the minimum window.
Example 3:

Input: s = "a", t = "aa"
Output: ""
Explanation: Both 'a's from t must be included in the window.
Since the largest window of s only has one 'a', return empty string.
*/

func minWindow(s string, t string) string {
	length := len(s)
	right := 0
	left := 0

	sHash := make(map[rune]int)
	tHash := make(map[rune]int)

	result := ""

	for _, v := range t {
		if value, ok := tHash[v]; ok {
			tHash[v] = value + 1
		} else {
			tHash[v] = 1
		}
	}

	for right <= length && left <= length {
		if check(sHash, tHash) {
			if result == "" || right-left < len(result) {
				result = s[left:right]
			}
			sHash[rune(s[left])] = sHash[rune(s[left])] - 1
			left++
		} else {
			if right < length {
				if count, ok := sHash[rune(s[right])]; ok {
					sHash[rune(s[right])] = count + 1
				} else {
					sHash[rune(s[right])] = 1
				}
			}
			right++
		}
	}

	return result
}

func check(sHash, tHash map[rune]int) bool {
	for key, tv := range tHash {
		sv, ok := sHash[key]
		if !ok {
			return false
		}
		if sv < tv {
			return false
		}
	}
	return true
}
